intermediary.py

```diff
@@ -67,94 +69,102 @@ class Direction(enum.Enum):
+        self.percentage = 0.0
 
-    def set_state(self, mode, intensity):
-        self.mode = mode
-        self.intensity = intensity
 
-class RouteStripState:
-    def __init__(self, colour, direction, stops):
-        self.colour = colour 
+class Route:
+    def __init__(self, line: Line, direction: Direction, stops):
+        self.line = line
         self.direction = direction
-        self.stops = stops
-        print(colour, 'has', len(stops))
-        self.stop_states = [StopState(colour) for _ in range(len(stops))]
+        self.stops: List[Stop] = [Stop(stop) for stop in stops]
 
-    def clear_stops(self):
-        for s in self.stop_states:
-            s.reset()
+        print(f'{line} has {len(stops)} stops')
 
-    def set_stop_state(self, stop_id, mode, intensity): 
-        print(f'Setting stop {stop_id} to mode {mode} ({intensity}) on {self.colour}, {self.direction}.')
-        i = self.stops.index(stop_id)
-        print(f' Stop index {i}')
+    def reset_stops(self):
+        for s in self.stops:
+            s.state = StopState.Empty
 
-        self.stop_states[i].set_state(mode, intensity)
-    
+    def get_stop_by_id(self, stop_id: str):
+        stop_index = [i for i, stop in enumerate(self.stops)
+                      if stop.name == stop_id]
+        if len(stop_index):
+            return stop_index[0]
+        return None
 
-class Intermediary:
-    def __init__(self):
-        self.strip_states = {}
-        for dest, line in dest_to_strip.items():
-            if line not in self.strip_states:
-                self.strip_states[line] = RouteStripState(line[0], line[1], route_stops['RouteStrip.'+line[0].name])
+    def set_stop_state(self, stop_id: str, state: StopState,
+                       percentage: float = 0):
+        stop_index = self.get_stop_by_id(stop_id)
 
-        # mapping of 2-tuple (line, direction) to a list of pin numbers.
-        self.used_strips = {}
+        if stop_index is None:
+            return
 
+        self.stops[stop_index].state = state
+        self.stops[stop_index].percentage = percentage
+
+
+class Intermediary:
+    def __init__(self):
         self.builder = message.MessageBuilder()
-        self.serial = device.SerialDevice('COM7', 9600)
         self.data = ''
 
+        self.line_strip_mapping = {}
+
+        self.routes = {}
+        for _, (line, direction) in dest_to_strip.items():
+            if (line, direction) not in self.routes:
+                self.routes[(line, direction)] = \
+                    Route(line, direction, route_stops['RouteStrip.' + line.name])
+
     @staticmethod
     def request_train_data():
         # TODO: connect to provider server
         return data_passer.get_train_data(('SPRP', ))
 
     @staticmethod
-    def split_line_name(name):
+    def split_line_name(name: str):
         return name.replace(' Line', '').replace(' line', '').split(' - ')
 
-    def update_state(self, interval):
+    def update_state(self, interval: int):
```

Comment:
Interval is actually unused. Was going to use it to compute percentage, but that's done in the server-side now.